nested columns + arrays = array columns!

clintropolis opened the pull request:
### Description
This PR adds support for storing arrays of primitive values (`ARRAY<STRING>`, `ARRAY<LONG>` and `ARRAY<DOUBLE>`) as specialized nested columns, instead of the current strategy which stores a separate nested column for each array element.

Basically it works by adding an additional value dictionary for all array values to our current `STRING` -> `LONG` -> `DOUBLE` dictionary stack. The array dictionary stores `int[]` values, where the elements are globalIds that point to values in the `STRING`/`LONG`/`DOUBLE` value dictionaries.

![E110E6BE-2856-4E09-BADA-575284BB3187 2](https://user-images.githubusercontent.com/1577461/218812357-45f333c2-389d-4ad0-bee0-0e8b7c33a1cf.png)

So instead of the 2-phase lookup that scalar nested columns have, nested array values have a 3-phase lookup - first lookup localId -> globalId, which gives an int[] of global ids which can then be translated into the actual values.

This dictionary is stored with a newly added `FrontCodedIntArrayIndexed`, which is pretty similar to `FrontCodedIndexed` added in #12277, but storing `int[]` instead of `byte[]`. There is probably more code that could be shared between the two implementations (currently there is none and `FrontCodedIntArrayIndexed` and its writer are complete copies).

_additional description TBD_

changes:
* add support for storing nested arrays of string, long, and double values as specialized nested columns instead of breaking them into separate element columns
* nested column typic mimic behavior means that columns ingested with only root arrays of primitive values will be ARRAY typed columns
* neat test stuff


<img width="1242" alt="Screenshot 2023-02-02 at 6 54 05 PM" src="https://user-images.githubusercontent.com/1577461/218803387-971251c3-879d-4856-9583-bce1f165f680.png">

<img width="995" alt="Screenshot 2023-02-07 at 2 21 33 PM" src="https://user-images.githubusercontent.com/1577461/218803657-af960ae6-c576-492c-8076-a9cd17e7e7e4.png">

<img width="772" alt="Screenshot 2023-02-07 at 5 05 21 PM" src="https://user-images.githubusercontent.com/1577461/218803814-9fd5edaf-2d23-4daa-9e0e-38cf17bd0b30.png">


#### Release note
<!-- Give your best effort to summarize your changes in a couple of sentences aimed toward Druid users. 

If your change doesn't have end user impact, you can skip this section.

For tips about how to write a good release note, see [Release notes](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#release-notes).

-->


<hr>


<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

clintropolis commented:
I want to get #13809 finished first and pull those changes into this PR to make things consistent with null value coercion

imply-cheddar commented on processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java:
```diff
@@ -480,11 +480,25 @@ public Expr apply(List<Expr> args)
+            @Nullable Object maybeArrayOfLiterals
+        )
+        {
+          ExprEval<?> eval = ExprEval.bestEffortOf(maybeArrayOfLiterals);
```
Would it make sense to break `ExprEval.bestEffortOf` into a bunch of checks for different groups of expected types (i.e. `ExprEval.maybeLiteral()` and `ExprEval.maybeArray`, etc.).  Calls to `bestEffortOf` can cascade through, but places like this that already know some of what they expect can call the one that more aligns with expectations?

clintropolis replied:
i split out `bestEffortArray(@Nullable List<?> theList)` and changed `StructuredDataProcessor` method to always pass a `List` when attempting to process arrays

imply-cheddar commented on processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java:
```diff
@@ -480,11 +480,25 @@ public Expr apply(List<Expr> args)
+        )
+        {
+          ExprEval<?> eval = ExprEval.bestEffortOf(maybeArrayOfLiterals);
+          if (eval.type().isArray() && eval.type().getElementType().isPrimitive()) {
+            return StructuredDataProcessor.ProcessedLiteral.NULL_LITERAL;
+          }
```
The if condition makes me believe that we found an array and should be joyously celebrating, but the return value is a `NULL_LITERAL`, why is it correct to return a `NULL_LITERAL` and how is that different from just returning `null`?

clintropolis replied:
see other comment about the contracts of the process literal methods. Here the json_paths expression doesn't care about the literal size estimations, so it just ignores them, since all it is really looking for is the `StructuredDataProcessor.ProcessResults` to get the set of paths. Returning non-null on the array method means that the processing on that path stops, so the path is to the array, not a separate path for each array element.

I'll try to explain a bit better with comments

clintropolis replied:
updated javadocs and comments on stuff so is hopefully clearer

imply-cheddar commented on processing/src/main/java/org/apache/druid/query/metadata/SegmentAnalyzer.java:
```diff
@@ -134,6 +134,11 @@ public Map<String, ColumnAnalysis> analyze(Segment segment)
           }
           break;
+        case ARRAY:
+          // todo (clint): this is wack, but works for now because arrays are always nested complex columns...
```
Instead of a todo like this.  How about an if check that validates that it's a nested column and, if it's not a nested column, throws an Exception that is clearly saying that the code is currently written assuming that the arrays are delivered only by nested columns?

clintropolis replied:
yep, i'm planning on replacing this todo with a 'what we need to do in the future', just haven't got to it yet

clintropolis replied:
split out into an `analyzeArrayColumn` that is a mostly empty stub

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -59,16 +62,43 @@ public class NestedDataColumnIndexer implements DimensionIndexer<StructuredData,
-        fieldIndexers.put(fieldName, fieldIndexer);
+      // null value is always added to the global dictionary as id 0, so we can ignore them here
+      if (fieldValue != null) {
+        // why not
```
Because I don't know "why?"  (if this comment is attempting to add information, more words please).

clintropolis replied:
oops, this needs removed

clintropolis replied:
removed stale comment

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -59,16 +62,43 @@ public class NestedDataColumnIndexer implements DimensionIndexer<StructuredData,
+        Object maybeArrayOfLiterals
+    )
+    {
+      final ExprEval<?> maybeLiteralArray = ExprEval.bestEffortOf(maybeArrayOfLiterals);
+      if (maybeLiteralArray.type().isArray() && maybeLiteralArray.type().getElementType().isPrimitive()) {
+        final String fieldName = NestedPathFinder.toNormalizedJsonPath(fieldPath);
+        LiteralFieldIndexer fieldIndexer = fieldIndexers.get(fieldName);
+        if (fieldIndexer == null) {
+          estimatedFieldKeySize += StructuredDataProcessor.estimateStringSize(fieldName);
+          fieldIndexer = new LiteralFieldIndexer(globalDictionary);
+          fieldIndexers.put(fieldName, fieldIndexer);
+        }
+        return fieldIndexer.processValue(maybeLiteralArray);
+      }
+      return null;
```
This looks a lot like code in NestedDataExpressions, except this doesn't return the `NULL_LITERAL`.  I find myself wondering if the NestedDataExpressions code shouldn't look like this?

clintropolis replied:
the contract of `processArrayOfLiteralsField` is supposed to return a `ProcessedLiteral` if and only if the value was an array of literals (it is marked `@Nullable`). `processLiteralField` is not nullable, and must always return a `ProcessedLiteral`.

The `StructuredDataProcessor` code when processing some input and it encounters arrays will first attempt to `processArrayOfLiteralField`, if it returns something, it was an array, else it must instead process the array elements recursively. `processLiteralField` is called on everything that isn't a map or array.

I'll see if i can clarify it better

clintropolis replied:
i modified this stuff a bit and updated javadocs so it is hopefully clearer, the new abstract method names are `processField` and `processArrayField` and the latter indicates that returning a non-null value halts further processing of arrays, otherwise the processor will continue for each element of the array

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -145,6 +175,10 @@ public DimensionSelector makeDimensionSelector(
     final int dimIndex = desc.getIndex();
     final ColumnValueSelector<?> rootLiteralSelector = getRootLiteralValueSelector(currEntry, dimIndex);
     if (rootLiteralSelector != null) {
+      final LiteralFieldIndexer root = fieldIndexers.get(NestedPathFinder.JSON_PATH_ROOT);
+      if (root.getTypes().getSingleType().isArray()) {
+        throw new UnsupportedOperationException("Not supported");
+      }
```
I'm reading this as "if all we have are root-level entries and they are always arrays, then throw a UOE exception".  I'm pretty sure I'm reading it wrong, but wishing the error message gave me more context without me feeling like I need to expand lines on the review to know what this is validating.

clintropolis replied:
heh, you are reading it correctly, currently the 'single typed root' dimension selector should only be used for scalar string columns, everything else should use the column value selector instead, will try to clarify with comments

imply-cheddar replied:
Or put more words into the exception message so it's clear what's not supported?

clintropolis replied:
clarified exception

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -411,7 +445,7 @@ public Object getObject()
           final StructuredData data = (StructuredData) dims[dimIndex];
           if (data != null) {
-            return data.getValue();
+            return ExprEval.bestEffortOf(data.getValue()).value();
```
Is this counting on coercion or something?

clintropolis replied:
yeah, i did this to make it consistent with the value that is stored in the dictionary (since the selector is used for merging/persisting)

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -437,41 +471,51 @@ public Class<?> classOfObject()
+        case ARRAY:
+          // skip empty arrays for now, they will always be called 'string' arrays, which isn't very helpful here since
+          // it will pollute the type set
+          Preconditions.checkNotNull(columnType.getElementType(), "Array element type must not be null");
```
If this were to ever happen, I think I'd want to know which field was the bad one.

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -437,41 +471,51 @@ public Class<?> classOfObject()
+          sizeEstimate = globalDimensionDictionary.addDoubleValue(eval.asDouble());
+          return new StructuredDataProcessor.ProcessedLiteral<>(eval.asDouble(), sizeEstimate);
+        case ARRAY:
+          // skip empty arrays for now, they will always be called 'string' arrays, which isn't very helpful here since
+          // it will pollute the type set
```
How does it do the skipping of empties?

clintropolis replied:
it was missing some code to do that, updated and added tests for nulls, empties, and arrays of nulls

imply-cheddar reviewed (commented):
Only got through a few of the files so far, but need to go and come back to it.

clintropolis commented on processing/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java:
```diff
@@ -77,22 +76,36 @@ public static InputRow parse(InputRowSchema inputRowSchema, Map<String, Object>
       LinkedHashSet<String> dimensions = new LinkedHashSet<>(dimensionsSpec.getDimensionNames());
-      dimensions.addAll(Sets.difference(rawInputRow.keySet(), dimensionsSpec.getDimensionExclusions()));
+      for (String field : rawInputRow.keySet()) {
+        if (timestampSpec.getTimestampColumn().equals(field) || dimensionsSpec.getDimensionExclusions().contains(field)) {
```
yeah, can adjust, though after an earlier discussion i just want to remove `getDimensions` from `InputRow`, but haven't decided if in this PR or some later change.

clintropolis commented on extensions-core/parquet-extensions/src/test/java/org/apache/druid/data/input/parquet/NestedColumnParquetReaderTest.java:
```diff
@@ -181,7 +181,7 @@ public void testNestedColumnSchemalessNestedTestFileNoNested() throws IOExceptio
 
     List<InputRow> rows = readAllRows(reader);
-    Assert.assertEquals(ImmutableList.of("dim1", "metric1", "timestamp"), rows.get(0).getDimensions());
+    Assert.assertEquals(ImmutableList.of("dim1", "metric1"), rows.get(0).getDimensions());
```
this is related to the change in dimension filter in `MapBasedInputRow` to always filter out the timestamp spec column.

In 'normal' production code the timestamp spec is added to dimension exclusions so the code in `MapBasedInputRow` that computes dimensions to ensure time would not be in the dimensions list. However, in test code, especially in processing, which doesn't have access to the methods that take a dataschema and transform it into an input row schema, its pretty easy to not explicitly add timestamp column to the dimensions exclusion list. So as a result of not manually adding timestamp column to exclusions, it would end up in the dimensions list in schema discovery modes, as a string (or nested column, depending on config), which when doing rollup tests means it ends up as part of the rollup key, and so on. (Again this doesn't happen in current production code because it goes through that translator utility method).

I made the change there to always filter the timestamp spec column from the dimensions list to make it easier to not write wrong tests for schema discovery mode, which caused the change here and other places.

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/nested/GlobalDimensionDictionary.java:
```diff
@@ -42,45 +47,107 @@
   private final ComparatorDimensionDictionary<String> stringDictionary;
   private final ComparatorDimensionDictionary<Long> longDictionary;
   private final ComparatorDimensionDictionary<Double> doubleDictionary;
+  private final Set<Object[]> stringArrays;
+  private final Set<Object[]> longArrays;
+  private final Set<Object[]> doubleArrays;
```
Why this separation by type?  Isn't is just an `array` and the ids are any other scalar thing in the global dictionary?

clintropolis replied:
this is also used for merging segments, so the original ints they referred to might have been remapped when the string/long/double dictionaries are merged beneath them, so we have to round trip like this to lookup the new values

clintropolis replied:
oh maybe this isn't used for merging segments... let me get back to you on this one

imply-cheddar commented on processing/src/test/resources/nested-array-test-data.json:
```diff
@@ -0,0 +1,14 @@
+{"timestamp": "2023-01-01T00:00:00", "arrayString": ["a", "b"],       "arrayStringNulls": ["a", "b"],         "arrayLong":[1, 2, 3],    "arrayLongNulls":[1, null,3],   "arrayDouble":[1.1, 2.2, 3.3],  "arrayDoubleNulls":[1.1, 2.2, null],  "arrayVariant":["a", 1, 2.2],     "arrayObject":[{"x": 1},{"x":2}]}
+{"timestamp": "2023-01-01T00:00:00", "arrayString": ["a", "b", "c"],  "arrayStringNulls": [null, "b"],        "arrayLong":[2, 3],                                       "arrayDouble":[3.3, 4.4, 5.5],  "arrayDoubleNulls":[999, null, 5.5],  "arrayVariant":[null, null, 2.2], "arrayObject":[{"x": 3},{"x":4}]}
+{"timestamp": "2023-01-01T00:00:00", "arrayString": ["b", "c"],       "arrayStringNulls": ["d", null, "b"],   "arrayLong":[1, 2, 3, 4], "arrayLongNulls":[1, 2, 3],     "arrayDouble":[1.1, 3.3],       "arrayDoubleNulls":[null, 2.2, null], "arrayVariant":[1, null, 1],      "arrayObject":[null,{"x":2}]}
+{"timestamp": "2023-01-01T00:00:00", "arrayString": ["d", "e"],       "arrayStringNulls": ["b", "b"],         "arrayLong":[1, 4],       "arrayLongNulls":[1],           "arrayDouble":[2.2, 3.3, 4.0],                                        "arrayVariant":["a", "b", "c"],   "arrayObject":[{"x": null},{"x":2}]}
+{"timestamp": "2023-01-01T00:00:00", "arrayString": null,                                                     "arrayLong":[1, 2, 3],    "arrayLongNulls":null,          "arrayDouble":[1.1, 2.2, 3.3],  "arrayDoubleNulls":null,                                                "arrayObject":[{"x": 1000},{"y":2000}]}
+{"timestamp": "2023-01-01T00:00:00", "arrayString": ["a", "b"],       "arrayStringNulls": null,                                         "arrayLongNulls":[null, 2, 9],  "arrayDouble":null,             "arrayDoubleNulls":[999, 5.5, null],  "arrayVariant":["a", 1, 2.2],     "arrayObject":[{"a": 1},{"b":2}]}
+{"timestamp": "2023-01-01T00:00:00",                                  "arrayStringNulls": ["a", "b"],         "arrayLong":null,         "arrayLongNulls":[2, 3],                                        "arrayDoubleNulls":[null, 1.1],       "arrayVariant":null,              "arrayObject":[{"x": 1},{"x":2}]}
+{"timestamp": "2023-01-02T00:00:00", "arrayString": ["a", "b"],       "arrayStringNulls": ["a", "b"],         "arrayLong":[1, 2, 3],    "arrayLongNulls":[1, null,3],   "arrayDouble":[1.1, 2.2, 3.3],  "arrayDoubleNulls":[1.1, 2.2, null],  "arrayVariant":["a", 1, 2.2],     "arrayObject":[{"x": 1},{"x":2}]}
+{"timestamp": "2023-01-02T00:00:00", "arrayString": ["a", "b", "c"],  "arrayStringNulls": [null, "b"],        "arrayLong":[2, 3],                                       "arrayDouble":[3.3, 4.4, 5.5],  "arrayDoubleNulls":[999, null, 5.5],  "arrayVariant":[null, null, 2.2], "arrayObject":[{"x": 3},{"x":4}]}
+{"timestamp": "2023-01-02T00:00:00", "arrayString": ["b", "c"],       "arrayStringNulls": ["d", null, "b"],   "arrayLong":[1, 2, 3, 4], "arrayLongNulls":[1, 2, 3],     "arrayDouble":[1.1, 3.3],       "arrayDoubleNulls":[null, 2.2, null], "arrayVariant":[1, null, 1],      "arrayObject":[null,{"x":2}]}
+{"timestamp": "2023-01-02T00:00:00", "arrayString": ["d", "e"],       "arrayStringNulls": ["b", "b"],         "arrayLong":[1, 4],       "arrayLongNulls":[1],           "arrayDouble":[2.2, 3.3, 4.0],                                        "arrayVariant":["a", "b", "c"],   "arrayObject":[{"x": null},{"x":2}]}
+{"timestamp": "2023-01-02T00:00:00", "arrayString": null,                                                     "arrayLong":[1, 2, 3],    "arrayLongNulls":null,          "arrayDouble":[1.1, 2.2, 3.3],  "arrayDoubleNulls":null,                                                "arrayObject":[{"x": 1000},{"y":2000}]}
+{"timestamp": "2023-01-02T00:00:00", "arrayString": ["a", "b"],       "arrayStringNulls": null,                                         "arrayLongNulls":[null, 2, 9],  "arrayDouble":null,             "arrayDoubleNulls":[999, 5.5, null],  "arrayVariant":["a", 1, 2.2],     "arrayObject":[{"a": 1},{"b":2}]}
+{"timestamp": "2023-01-02T00:00:00",                                  "arrayStringNulls": ["a", "b"],         "arrayLong":null,         "arrayLongNulls":[2, 3],                                        "arrayDoubleNulls":[null, 1.1],       "arrayVariant":null,              "arrayObject":[{"x": 1},{"x":2}]}
```
What's the expectation of how arrays of arrays are handled?

clintropolis replied:
i'll add some tests for this, but assuming something like ARRAY<ARRAY<STRING>> there would be an array column for each outer array element

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/nested/NestedDataColumnSerializer.java:
```diff
@@ -104,6 +108,32 @@ public StructuredDataProcessor.ProcessedLiteral<?> processLiteralField(ArrayList
+            return StructuredDataProcessor.ProcessedLiteral.NULL_LITERAL;
+          }
+          catch (IOException e) {
+            throw new RuntimeException(":(");
```
This is not only throwing away the IOException, it's not even attempting to generate a nice message.  `:(` :P

clintropolis replied:
oops yes, this was a placeholder

clintropolis replied:
fixed

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/nested/NestedDataColumnSerializer.java:
```diff
@@ -215,14 +263,24 @@ public void serializeFields(SortedMap<String, NestedLiteralTypeInfo.MutableTypeS
+              globalDictionaryIdLookup
+          );
+        } else {
+          throw new IllegalArgumentException("wtf");
```
How about make it a `UOE` and include the type that caused us to get here :)

clintropolis replied:
oops this one was also a placeholder

clintropolis replied:
fixed

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/nested/NestedDataColumnSerializer.java:
```diff
@@ -251,33 +313,78 @@ public void serializeStringDictionary(Iterable<String> dictionaryValues) throws
+    doubleDictionarySerialized = true;
+  }
+
+  public void serializeArrayDictionary(Iterable<int[]> dictionaryValues) throws IOException
+  {
+    if (!stringDictionarySerialized) {
+      throw new ISE("Must serialize string value dictionary before serializing array dictionary for column [%s]", name);
+    }
+    if (!longDictionarySerialized) {
+      throw new ISE("Must serialize long value dictionary before serializing array dictionary for column [%s]", name);
+    }
+    if (!doubleDictionarySerialized) {
+      throw new ISE("Must serialize double value dictionary before serializing array dictionary for column [%s]", name);
+    }
+    if (arrayDictionarySerialized) {
+      throw new ISE("Array dictionary already serialized for column [%s], cannot serialize again", name);
+    }
```
It seems weird to me to have a bunch of public methods that throw a bunch of exceptions if they aren't called in the right order.  Why do each of them need to be exposed as public individually if they can only be called once in a specific order?  Why not have a single `serializeDictionaries` method that is called publicly which does the things in the correct order?

clintropolis replied:
yeah i agree it could use refactored, i added the exceptions to at least enforce this is done correctly because it was previously true as well just nothing was checking it, will look into making some adjustments

clintropolis replied:
have added `serializedDictionaries`

imply-cheddar commented on processing/src/test/java/org/apache/druid/query/aggregation/AggregationTestHelper.java:
```diff
@@ -509,50 +507,6 @@ public void createIndex(
-      String aggregators,
-      File outDir,
-      long minTimestamp,
-      Granularity gran,
-      int maxRowCount,
-      boolean rollup
-  ) throws Exception
-  {
-    try {
-      StringInputRowParser parser = mapper.readValue(parserJson, StringInputRowParser.class);
-      TransformSpec transformSpec;
-      if (transformSpecJson != null) {
-        transformSpec = mapper.readValue(transformSpecJson, TransformSpec.class);
-        parser = new TransformingStringInputRowParser(parser.getParseSpec(), parser.getEncoding(), transformSpec);
-      }
-
-      LineIterator iter = IOUtils.lineIterator(inputDataStream, "UTF-8");
-      List<AggregatorFactory> aggregatorSpecs = mapper.readValue(
-          aggregators,
-          new TypeReference<List<AggregatorFactory>>()
-          {
-          }
-      );
-
-      createIndex(
-          iter,
-          parser,
-          aggregatorSpecs.toArray(new AggregatorFactory[0]),
-          outDir,
-          minTimestamp,
-          gran,
-          true,
-          maxRowCount,
-          rollup
-      );
-    }
-    finally {
-      Closeables.close(inputDataStream, true);
-    }
-  }
```
Removing this method could quite easily break extensions.  IIRC, the `AggregationTestHelper` exists primarily to aid in building tests for AggregatorFactory extensions.  Did the same logic move somewhere else? If so, can we keep this method (maybe deprecated?) and have it call that other place?

clintropolis replied:
Hmm, I added this one in the PR that introduced nested columns. The same functionality that many of the methods that create segments in `AggregationTestHelper ` have is done better in `IndexBuilder`, which is why I focused on improving it and switching nested column tests to using it.

There were no users of this method after i refactored my tests so I removed it. I can always add it back if you're worried about it, but imo everyone should use `IndexBuilder` and we should deprecate most of `AggregationTestHelper`, at least the parts around making segments. I didn't not migrate the other users away from it in this PR though.

imply-cheddar commented on processing/src/test/java/org/apache/druid/segment/IndexBuilder.java:
```diff
@@ -157,27 +241,71 @@ public QueryableIndex buildMMappedIndex()
+      Preconditions.checkNotNull(inputFormat, "inputFormat");
+      Preconditions.checkNotNull(inputSourceTmpDir, "inputSourceTmpDir");
+
+      TransformSpec tranformer = transformSpec != null ? transformSpec : TransformSpec.NONE;
```
spell-check: `transformer`

imply-cheddar commented on processing/src/test/java/org/apache/druid/segment/IndexBuilder.java:
```diff
@@ -267,4 +395,44 @@ private static IncrementalIndex buildIncrementalIndexWithRows(
+        .setIndexSchema(schema)
+        .setMaxRowCount(maxRows)
+        .build();
+    TransformSpec tranformer = transformSpec != null ? transformSpec : TransformSpec.NONE;
```
spell-check

imply-cheddar commented on processing/src/test/java/org/apache/druid/segment/IndexBuilder.java:
```diff
@@ -267,4 +395,44 @@ private static IncrementalIndex buildIncrementalIndexWithRows(
+    InputSourceReader transformingReader = tranformer.decorate(reader);
+    try (CloseableIterator<InputRow> rowIterator = transformingReader.read()) {
+      while (rowIterator.hasNext()) {
+        incrementalIndex.add(rowIterator.next());
```
When `maxRows` is hit, are we expecting an exception?  Generally speaking, setting the `maxRows` on the tests is done as a way to force running queries against multiple segments, so I had expected to see a check for the numRows and incremental persists in anything that takes `maxRows`.

clintropolis replied:
Hmm, good question, this looks wired up to the `maxRowCount` but callers are not checking `canAppendRow` or anything like that and `buildIncrementalIndex` can only return a single `IncrementalIndex` so I think all it can do is explode. There is also `intermediatePersistSize` which can be used to force a bunch of incremental indexes to be written when `mergeIndexes` is called to make sure that segment merging happens.

Nothing much seems to be explicitly setting either of these things, and maybe could be removed or reworked in a follow-up.

imply-cheddar commented on sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/ArrayContainsOperatorConversion.java:
```diff
@@ -95,7 +95,7 @@ public DimFilter toDruidFilter(
     final DruidExpression rightExpr = druidExpressions.get(1);
 
-    if (leftExpr.isSimpleExtraction()) {
+    if (leftExpr.isSimpleExtraction() && !(leftExpr.getDruidType() != null && leftExpr.getDruidType().isArray())) {
```
What if it's an `array_contains()` over just a normal single-valued `String` column?  Shouldn't that also match the filter, pretending that each row contains an array of size 1?

clintropolis replied:
the type shouldn't show up as array here if it isn't an array column

imply-cheddar commented on sql/src/test/java/org/apache/druid/sql/calcite/BaseCalciteQueryTest.java:
```diff
@@ -705,6 +705,23 @@ public void testQuery(
         .run();
   }
 
+  public void testQuery(
+      final String sql,
+      final Map<String, Object> queryContext,
+      final List<Query<?>> expectedQueries,
+      final List<Object[]> expectedResults,
+      final RowSignature expectedResultSignature
+  )
+  {
+    testBuilder()
+        .sql(sql)
+        .queryContext(queryContext)
+        .expectedQueries(expectedQueries)
+        .expectedResults(expectedResults)
+        .expectedSignature(expectedResultSignature)
+        .run();
+  }
```
Part of me wonders if, instead of adding more functions with extra arguments to their signatures, we should switch the call-sites to using the builder directly?

clintropolis replied:
removed in favor of just calling `testBuilder`, this should probably just be done everywhere, but its kind of tedious to switch

clintropolis commented on processing/src/main/java/org/apache/druid/segment/data/FrontCodedIntArrayIndexed.java:
```diff
@@ -0,0 +1,524 @@
+    }
+    Indexed.checkIndex(index, adjustedNumValues);
+
+    // due to vbyte encoding, the null value is not actually stored in the bucket (no negative values), so we adjust
+    // the index
```
null is stored as a flag outside of the bucket because there is no way to store inside (we would typically use a length of -1 to indicate a null value, but since we cannot represent negative numbers it uses an external flag and index shifting. 0 length value is an empty array here and so separate from null.

imply-cheddar replied:
If I understood correctly, can I suggest updating the comment to be

```
// For arrays, there is a conundrum of how to encode null different from the empty array.  In this code, 
// we take the following approach:
//   1) the 0 entry in our dictionary is assumed to be indicative of a completely null entry
//   2) The 1 value in our dictionary is assumed to be the empty array
//   3) Instead of storing the 0 value in this dictionary, we employ an indexing-shifting technique, where the
//       dictionary never stores null, and starts with the empty array (index 1), but shifted by 1 such that what
//       is persisted on disk is actually 0 instead of 1.
// adjustIndex represents whether the null existed and thus whether we actually need to adjust the value
```

clintropolis replied:
this is like the same code/comments as in `FrontCodedIndexed`, just adapted for `int[]` instead of `byte[]`, what is the empty array here is the empty string there, so it is the same problem. I guess i can update the comments/javadocs on both of them to make it more apparent of why we special handle null values and how it relates to `VByte` encoding

clintropolis commented on processing/src/main/java/org/apache/druid/segment/UnnestStorageAdapter.java:
```diff
@@ -101,7 +101,7 @@ public Sequence<Cursor> makeCursors(
           ColumnCapabilities capabilities = cursor.getColumnSelectorFactory().getColumnCapabilities(dimensionToUnnest);
           if (capabilities != null) {
-            if (capabilities.isDictionaryEncoded().and(capabilities.areDictionaryValuesUnique()).isTrue()) {
+            if (!capabilities.isArray() && capabilities.isDictionaryEncoded().and(capabilities.areDictionaryValuesUnique()).isTrue()) {
```
so, this array column is dictionary encoded, but is not dictionary encoded in the correct way for the unnest dimension cursor, instead the dictionary ids stored in the column are the entire arrays id. Unnest will try to incorrectly use a dimension selector for these array columns.

Are there things legitimately typed as ARRAY but should use a dimension selector?

clintropolis replied:
i confirmed that this doesn't impact `MV_TO_ARRAY` when used in SQL because the planner unwraps it leaving just a direct column access for the underlying multi-value column, so the capabilities are `STRING` typed with `hasMultipleValues` set to true.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/nested/ArrayOfLiteralsFieldColumnWriter.java:
```diff
@@ -0,0 +1,85 @@
+  @Override
+  int[] processValue(int row, Object value)
+  {
+    if (value instanceof Object[]) {
```
it doesn't have to because `ExprEval` homogenizes the array types to `Object[]`

clintropolis commented on processing/src/main/java/org/apache/druid/segment/nested/CompressedNestedDataComplexColumn.java:
```diff
@@ -327,17 +390,58 @@ public void close()
+        if (elementNumber < 0) {
+          throw new IAE("Cannot make array element selector for path [%s], negative array index not supported for this selector", path);
+        }
+        DictionaryEncodedColumn<?> col = (DictionaryEncodedColumn<?>) getColumnHolder(arrayField).getColumn();
+        ColumnValueSelector<?> arraySelector = col.makeColumnValueSelector(readableOffset);
+        return new BaseSingleValueDimensionSelector()
+        {
+          @Nullable
+          @Override
+          protected String getValue()
+          {
+            Object o = arraySelector.getObject();
+            if (o instanceof Object[]) {
+              Object[] array = (Object[]) o;
+              if (elementNumber < array.length) {
+                Object element = array[elementNumber];
+                if (element == null) {
+                  return null;
+                }
+                return String.valueOf(element);
+              }
+            }
+            return null;
+          }
+
+          @Override
+          public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+          {
+            arraySelector.inspectRuntimeShape(inspector);
+          }
+        };
```
the dictionary ids here are complete arrays, not the elements of the array, dimension selector doesn't really work for this because lookupName returns a string for a dictionary id instead of ARRAY<STRING>/ARRAY<LONG>/ARRAY<DOUBLE> array columns we have

clintropolis commented on processing/src/main/java/org/apache/druid/segment/nested/CompressedNestedDataComplexColumn.java:
```diff
@@ -380,9 +551,61 @@ public VectorObjectSelector makeVectorObjectSelector(List<NestedPathPart> path,
+          public Object[] getObjectVector()
+          {
+            if (readableOffset.getId() != id) {
+              final Object[] delegate = arraySelector.getObjectVector();
+              for (int i = 0; i < arraySelector.getCurrentVectorSize(); i++) {
```
agree this isn't very optimized, i was planning to introduce some form of array selector in a future PR to optimize selecting individual elements

clintropolis commented on processing/src/main/java/org/apache/druid/segment/nested/CompressedNestedDataComplexColumn.java:
```diff
@@ -426,6 +765,12 @@ public ColumnIndexSupplier getColumnIndexSupplier(List<NestedPathPart> path)
     if (fields.indexOf(field) < 0) {
+      if (!path.isEmpty() && path.get(path.size() - 1) instanceof NestedPathArrayElement) {
+        final String arrayField = getField(path.subList(0, path.size() - 1));
+        if (fields.indexOf(arrayField) >= 0) {
+          return NoIndexesColumnIndexSupplier.getInstance();
+        }
```
no, we are writing out some value indexes for both whole arrays and individual elements, I just haven't wired everything up to native filters yet to take advantage of them so i'm forcing it to use value matchers until then so they work correctly

clintropolis commented on processing/src/main/java/org/apache/druid/segment/nested/GlobalDictionaryEncodedFieldColumnWriter.java:
```diff
@@ -231,9 +252,16 @@ public void writeTo(int finalRowCount, FileSmoosher smoosher) throws IOException
       @Override
       public long getSerializedSize() throws IOException
       {
+        final long arraySize;
+        if (arrayElements.size() > 0) {
+          arraySize = arrayElementDictionaryWriter.getSerializedSize() + arrayElementIndexWriter.getSerializedSize();
+        } else {
+          arraySize = 0;
+        }
```
yeah i've got a whole pile of stats i'd like to track but haven't got to yet, will add this one to the list.

clintropolis commented:
I've fixed the behavior of `UNNEST` with regards to how it handles `null` and `[]` values for `ARRAY` typed inputs. Per experiments with postgres, `null` and empty arrays `[]` are supposed to output nothing when fed into `UNNEST`:

<img width="1259" alt="Screenshot 2023-03-23 at 2 02 11 AM" src="https://user-images.githubusercontent.com/1577461/227349056-a3d6794c-dcba-4222-86ea-ecd9ceec65f5.png">

<img width="1015" alt="Screenshot 2023-03-23 at 2 04 02 AM" src="https://user-images.githubusercontent.com/1577461/227349099-04e9c280-7ff4-4388-9e37-0708dad62cf4.png">

<img width="1270" alt="Screenshot 2023-03-23 at 2 04 57 AM" src="https://user-images.githubusercontent.com/1577461/227349136-a7b5a2d5-ce81-4ef4-bfb2-42d6bb552687.png">

 but currently we are treating them as `[null]`, which is probably unfortunately what we have to keep doing for multi-value `STRING` types because we have trouble distinguishing `null`, `[]` and `[null]`, but no need to do this with actual arrays. So, I have fixed `UnnestColumnValueSelectorCursor` to consume `null` and `[]` values, fixing the behavior and added some tests.

clintropolis commented on processing/src/main/java/org/apache/druid/data/input/impl/DelimitedInputFormat.java:
```diff
@@ -41,6 +42,18 @@
+        false,
+        0
+    );
+  }
```
removed since there was only one caller

clintropolis commented on processing/src/main/java/org/apache/druid/data/input/impl/JsonInputFormat.java:
```diff
@@ -41,6 +41,14 @@ public class JsonInputFormat extends NestedInputFormat
+      null,
+      null
+  );
+
```
removed

clintropolis commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java:
```diff
@@ -86,93 +94,123 @@ public NestedDataColumnMerger(
+        boolean allNulls = dimValues == null || allNull(dimValues.getSortedStrings()) &&
+                                                allNull(dimValues.getSortedLongs()) &&
+                                                allNull(dimValues.getSortedDoubles()) &&
+                                                dimValues.getArrayCardinality() == 0;
```
moved to `GlobalDictionarySortedCollector`

clintropolis commented on processing/src/main/java/org/apache/druid/segment/nested/CompressedNestedDataComplexColumn.java:
```diff
@@ -327,17 +390,58 @@ public void close()
+    if (!path.isEmpty() && path.get(path.size() - 1) instanceof NestedPathArrayElement) {
+      final NestedPathPart lastPath = path.get(path.size() - 1);
+      final String arrayField = getField(path.subList(0, path.size() - 1));
+      if (fields.indexOf(arrayField) >= 0) {
```
refactored to use `fieldIndex` after looking up once

clintropolis commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java:
```diff
@@ -234,7 +272,9 @@ private GlobalDictionarySortedCollector getSortedIndexFromV1QueryableAdapterNest
         column.getLongDictionary(),
-        column.getDoubleDictionary()
+        column.getDoubleDictionary(),
+        column.getArraysIterable(),
+        column.getArrayDictionary().size()
```
it felt strange to try to wire this iterable up to wrap in an `Indexed` when all we need it for is to iterate over

clintropolis commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java:
```diff
@@ -280,4 +320,122 @@ private <T> boolean allNull(Indexed<T> dimValues)
+      }
+      final int[] newIdsWhoDis = new int[next.length];
+      for (int i = 0; i < next.length; i++) {
+        if (next[i] == null) {
+          newIdsWhoDis[i] = 0;
+        } else if (next[i] instanceof String) {
+          newIdsWhoDis[i] = idLookup.lookupString((String) next[i]);
+        } else if (next[i] instanceof Long) {
+          newIdsWhoDis[i] = idLookup.lookupLong((Long) next[i]);
+        } else if (next[i] instanceof Double) {
+          newIdsWhoDis[i] = idLookup.lookupDouble((Double) next[i]);
+        } else {
+          newIdsWhoDis[i] = -1;
+        }
+        Preconditions.checkArgument(
+            newIdsWhoDis[i] >= 0,
+            "unknown global id [%s] for value [%s]",
+            newIdsWhoDis[i],
+            next[i]
+        );
```
So, we are dealing with them here like this so we can lookup the new global id from the newly merged lower scalar value dictionaries. Otherwise we would need the mappings of old ids to new ids, which we don't currently have anywhere, and its a lot more complicated since its per segment. This way we just lookup the old values and after the lower dictionaries are merged, just lookup the array elements for the newly sorted values

clintropolis replied:
i think since this stuff is all working i'd like to save further changes/optimizations for a follow-up

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/nested/NestedDataColumnSerializer.java:
```diff
@@ -81,46 +87,75 @@ public class NestedDataColumnSerializer implements GenericColumnSerializer<Struc
+            return ProcessedValue.NULL_LITERAL;
+          }
+          catch (IOException e) {
+            throw new RE(e, "Failed to write field [%s] value [%s]", fieldPath, array);
```
This exception leaks the content of the data if it ever gets built.  I think the "best" we can do here is mention which field it was trying to write.

clintropolis replied:
yeah, that's fair, i can change it, was thinking in terms of helping myself debug what wasn't handled correctly but it is leaky

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/nested/NestedFieldTypeInfo.java:
```diff
@@ -31,29 +32,36 @@
   private static final byte DOUBLE_MASK = 1 << 3;
 
-  public static NestedLiteralTypeInfo read(ByteBuffer buffer, int length)
+  private static final byte STRING_ARRAY_MASK = 1 << 4;
+
+  private static final byte LONG_ARRAY_MASK = 1 << 5;
+
+  private static final byte DOUBLE_ARRAY_MASK = 1 << 6;
```
I dunno if I'm being stingy, but couldn't we have an `array` mask that is applied and then we check the type mask after that?  Would mean that we can add array support to this and only consume 1 more bit instead of consuming 3 more bits.

clintropolis replied:
that would work if we only set 1 type, or if any row being an array meant all rows are arrays, but it doesn't currently, so we sort of need to know what type of array, at least the way things currently work

imply-cheddar replied:
what about composite arrays?  Like if the array has a String and a double?  Or a String and an Object?

I had intepretted these as "this type exists in the column" not as "every row is this type".  In which case, if there is an array, then it is true that the array type exists in the column, no?  And then if the only scalar type is a `long`, then you would know that the arrays are all longs, right?

imply-cheddar commented on sql/src/test/java/org/apache/druid/sql/calcite/BaseCalciteQueryTest.java:
```diff
@@ -190,6 +190,7 @@ public static void setupNullValues()
 
   public static final Map<String, Object> QUERY_CONTEXT_NO_STRINGIFY_ARRAY =
       DEFAULT_QUERY_CONTEXT_BUILDER.put(QueryContexts.CTX_SQL_STRINGIFY_ARRAYS, false)
+                                   .put(PlannerContext.CTX_ENABLE_UNNEST, true)
```
Not sure how much I care about this, but I expected this to be put on the Unnest/Array tests instead of across all of the tests in this base class.

clintropolis replied:
`QUERY_CONTEXT_NO_STRINGIFY_ARRAY` is afaik only used on `CalciteArrayQueryTest` and `CalciteNestedDataQueryTest` so i think this is fine

clintropolis commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -426,55 +471,71 @@ public Class<?> classOfObject()
+              sizeEstimate = globalDimensionDictionary.addStringArray(stringArray);
+              return new StructuredDataProcessor.ProcessedValue<>(stringArray, sizeEstimate);
+            default:
+              throw new IAE("Unhandled type: %s", columnType);
```
We don't really expect it to happen because the `StructuredDataProcessor` should have done its job correctly and only defined fields for primitive and array of primitive, so this is more of a sanity check

imply-cheddar replied:
Ah, so `array-of-arrays` should be handled in `StructuredDataProcessor` instead?

Even after this PR, the handling of `array-of-arrays` and `array-of-objects` is still a bit muddy, right?  (it's falling back to the "an array is just an object with numbers for field names" behavior?)

clintropolis commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -426,55 +471,71 @@ public Class<?> classOfObject()
+              throw new IAE("Unhandled type: %s", columnType);
+          }
+        case STRING:
+        default:
```
I don't think we should hit the default case, i'm not entirely sure why i wrote it like this originally

clintropolis commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnIndexer.java:
```diff
@@ -426,55 +471,71 @@ public Class<?> classOfObject()
+          return new StructuredDataProcessor.ProcessedValue<>(eval.asDouble(), sizeEstimate);
+        case ARRAY:
+          // sanity check, this should never happen
+          Preconditions.checkNotNull(
+              columnType.getElementType(),
+              "Array type [%s] for value [%s] missing element type, how did this possibly happen?",
+              eval.type(),
+              eval.valueOrDefault()
+          );
```
this shouldn't ever happen, but I guess i can wrap it in an if

imply-cheddar replied:
Just for clarity, I was suggestion that instead of using `Preconditions` you use an if statement instead.

clintropolis commented on processing/src/main/java/org/apache/druid/math/expr/ExprEval.java:
```diff
@@ -370,6 +370,18 @@ public static ExprEval ofComplex(ExpressionType outputType, @Nullable Object val
+    // empty
+    NonnullPair<ExpressionType, Object[]> coerced = coerceListToArray(theList, false);
+    if (coerced == null) {
+      return bestEffortOf(null);
```
I think it would probably be best to have null handled first in `bestEffortOf` so that if we ever decide to represent `null` as something more sensible (like introduce a null type), we won't be making a 'string' here, not to mention the saving of not running through a bunch of checks for a null value.

imply-cheddar replied:
Sure, I expected you to actually to create a method for `makeNullNode()` or something in `ExprEval` instead of newing it up directly here.  Just moving around the ordering of the statements is 🤷 .  Though, checking for `null` first is probably better than last.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/NestedDataColumnMerger.java:
```diff
@@ -86,99 +94,129 @@ public NestedDataColumnMerger(
-        cardinality,
-        System.currentTimeMillis() - dimStartTime
-    );
+    catch (Throwable ioe) {
```
hmm, I don't remember why I added this, I think I was debugging something and needed an easy to place to catch to tell me what was failing for which column, and catching here seemed the best way to tell me what was messed up, but it could probably narrow it down to `IOException` and just rethrow it
